### PR TITLE
PTT buttons integration

### PR DESCRIPTION
**Pull request by zello-eugene**
This change set adds the ability to open the PTT Buttons screen, disable Bluetooth PTT buttons notifications in the ZelloWork app, and to intercept Bluetooth PTT buttons state changes.

### CODE REVIEW

File: zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java [hunk 1]
```diff
@@ -218,147 +219,147 @@ public void onMicrophonePermissionNotGranted() {
...
+
+	private void chooseStatus() {
+		AlertDialog.Builder builder = new AlertDialog.Builder(this);
+		Resources res = getResources();
+		Zello.getInstance().getAppState(_appState);
+
+		com.zello.sdk.Status status = _appState.getStatus();
+		int selection = -1;
+		switch (status) {
+			case AVAILABLE: {
```
Comment (adambailey-) [inline zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java]: braces for each case not necessary

File: zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java [hunk 2]
```diff
@@ -218,147 +219,147 @@ public void onMicrophonePermissionNotGranted() {
...
+		if (_appState.getStatusMessage() != null && !_appState.getStatusMessage().equals("")) {
+			selection = 3;
+		}
+
+		String[] items = new String[]{res.getString(R.string.menu_available), res.getString(R.string.menu_solo), res.getString(R.string.menu_busy), res.getString(R.string.menu_set_status_message)};
+		builder.setSingleChoiceItems(items, selection, new DialogInterface.OnClickListener() {
+			@Override
+			public void onClick(DialogInterface dialog, int which) {
+				switch (which) {
+					case 0: {
```
Comment (adambailey-) [inline zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java]: braces for each case not necessary

File: zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java [hunk 3]
```diff
@@ -218,147 +219,147 @@ public void onMicrophonePermissionNotGranted() {
...
+		builder.setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
+			@Override
+			public void onClick(DialogInterface dialog, int which) {
+				Zello.getInstance().setStatusMessage(statusEditText.getText().toString());
+			}
+		});
+		builder.setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
+			@Override
+			public void onClick(DialogInterface dialog, int which) {
+
```
Comment (adambailey-) [inline zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java]: Is an empty impl correct? Seems like it should dismiss the dialog

File: zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java [hunk 4]
```diff
@@ -11,233 +11,240 @@
...
 		}
+		return true;
+	}
 
-        return true;
-    }
+	@Override
+	public boolean onOptionsItemSelected(MenuItem item) {
+		switch (item.getItemId()) {
+			case R.id.menu_mute_contact: {
```
Comment (adambailey-) [inline zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java]: braces for each case not necessary

File: zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java [hunk 5]
```diff
@@ -11,233 +11,240 @@
...
+
+				return false;
+			}
+		});
+
+		// Audio modes
+		_speakerButton.setOnClickListener(new View.OnClickListener() {
+			@Override
+			public void onClick(View v) {
+				if (_audio != null) {
```
Comment (adambailey-) [inline zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java]: Minor, but if you could early return all of these OnClickListeners that would be good

File: zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java [hunk 6]
```diff
@@ -248,186 +255,206 @@ public void onMicrophonePermissionNotGranted() {
...
+	private void updateConnectChannelButton() {
+		boolean showConnect = false, connected = false, canConnect = false;
+
+		com.zello.sdk.ContactType type = _selectedContact.getType();
+		com.zello.sdk.ContactStatus status = _selectedContact.getStatus();
+		switch (type) {
+			case USER:
+			case GATEWAY:
+			case CHANNEL:
+			case GROUP: {
```
Comment (adambailey-) [inline zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java]: braces not necessary

File: zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java [hunk 7]
```diff
@@ -248,186 +255,206 @@ public void onMicrophonePermissionNotGranted() {
...
+		// Since Zello was initialized in the Activity, pass in this as Activity parameter
+		Zello.getInstance().selectContact(title, tabs, tab, theme, this);
+	}
+
+	private boolean getCanTalk() {
+		com.zello.sdk.ContactType type = _selectedContact.getType();
+		com.zello.sdk.ContactStatus status = _selectedContact.getStatus();
+		switch (type) {
+			case USER:
+			case GATEWAY: {
```
Comment (adambailey-) [inline zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java]: braces not necessary

File: zello-sdk-sample-signin/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -1,8 +1,15 @@
...
+    <string name="network">Network URL</string>
+    <string name="sign_in">Sign in"</string>
+    <string name="sign_out">Sign out"</string>
+    <string name="perishable">Do not save account data</string>
+    <string name="cancel">Cancel</string>
 
-    <string name="signing_in">SIGNING IN\u2026"</string>
-    <string name="signed_in">YOU ARE SIGNED IN.</string>
+    <string name="signing_in">Signing in\u2026"</string>
+    <string name="signed_in">You re signed in.</string>
```
Comment (adambailey-) [inline zello-sdk-sample-signin/src/main/res/values/strings.xml]: Looks like you're missing the '

File: zello-sdk-sample/build.gradle [hunk 9]
```diff
@@ -10,48 +11,31 @@ buildscript {
...
 }
+
+dependencies {
+    implementation fileTree(include: '*.jar', dir: 'libs')
+    implementation 'com.android.support:appcompat-v7:26.1.0'
+}
+
+repositories {
+    google()
+}
```
Comment (adambailey-) [inline zello-sdk-sample/build.gradle]: add newline

File: zello-sdk-sample/src/main/java/com/zello/sdk/sample/Helper.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+import android.content.Context;
+import android.content.SharedPreferences;
+
+public class Helper {
+
+	private static final String _loginPref = "login";
+
+	public static void saveValue(Context context, String key, String data) {
+		if (context != null && key != null && key.length() >= 0) {
```
Comment (adambailey-) [inline zello-sdk-sample/src/main/java/com/zello/sdk/sample/Helper.java]: Could you modify these methods to early return?

File: zello-sdk-sample/src/main/java/com/zello/sdk/sample/TalkActivity.java [hunk 11]
```diff
@@ -63,48 +78,65 @@
 	protected void onCreate(Bundle savedInstanceState) {
 		super.onCreate(savedInstanceState);
 
+		final Object event = new Object();
```
Comment (adambailey-) [inline zello-sdk-sample/src/main/java/com/zello/sdk/sample/TalkActivity.java]: Please delete all of this!

File: zello-sdk/src/com/zello/sdk/Sdk.java [hunk 12]
```diff
@@ -557,14 +574,34 @@ void openMainScreen() {
...
 		}
 	}
 
 	//endregion
 
+	//region App settings
+
+	void showPttButtonsScreen(Activity activity) {
+		Context context = activity != null ? activity : _context;
+		if (context != null) {
```
Comment (adambailey-) [inline zello-sdk/src/com/zello/sdk/Sdk.java]: early return this

File: zello-sdk/src/com/zello/sdk/Sdk.java [hunk 13]
```diff
@@ -965,24 +1038,38 @@ private void updateSelectedContact(Intent intent) {
...
 					event.onMicrophonePermissionNotGranted();
 				}
 			}
+	}
+
+	private void handleBtAccessoryState(Intent intent) {
+		if (intent == null) {
+			return;
 		}
+			BluetoothAccessoryType type = intToBtAccessoryType(intent.getIntExtra(Constants.EXTRA_TYPE, BluetoothAccessoryType.SPP.ordinal()));
```
Comment (adambailey-) [inline zello-sdk/src/com/zello/sdk/Sdk.java]: This looks improperly indented

File: zello-sdk/src/com/zello/sdk/Sdk.java [hunk 14]
```diff
@@ -1102,20 +1189,20 @@ private static String tabToString(Tab tab) {
 	}
 
 	private static String tabsToString(Tab[] tabs) {
-		String s = null;
+		StringBuilder s = null;
 		if (tabs != null) {
```
Comment (adambailey-) [inline zello-sdk/src/com/zello/sdk/Sdk.java]: Can you early return this please

File: zello-sdk/src/com/zello/sdk/Sdk.java [hunk 15]
```diff
@@ -1128,17 +1215,35 @@ private static Tab stringToTab(String s) {
...
+			return BluetoothAccessoryState.CONNECTED;
+		}
+		if (state == BluetoothAccessoryState.DISCONNECTED.ordinal()) {
+			return BluetoothAccessoryState.DISCONNECTED;
+		}
+		return BluetoothAccessoryState.ERROR;
+	}
+
 	private static String bytesToHex(byte[] data) {
 		if (data != null) {
```
Comment (adambailey-) [inline zello-sdk/src/com/zello/sdk/Sdk.java]: Early return

File: zello-sdk-sample/src/main/java/com/zello/sdk/sample/TalkActivity.java (hunk 11 above)
Comment (agatx) [reply zello-sdk-sample/src/main/java/com/zello/sdk/sample/TalkActivity.java]: LOL

File: zello-sdk/src/com/zello/sdk/Sdk.java [hunk 16]
```diff
@@ -81,33 +84,35 @@ void onCreate(String packageName, Context context) {
...
+					if (intent == null || _handler == null) {
+						return;
+					}
+					String action = intent.getAction();
+					if (action == null) {
+						return;
+					}
+					if (action.equals(Intent.ACTION_EXTERNAL_APPLICATIONS_AVAILABLE) || action.equals(Intent.ACTION_EXTERNAL_APPLICATIONS_UNAVAILABLE)) {
+						String[] pkgs = intent.getStringArrayExtra(Intent.EXTRA_CHANGED_PACKAGE_LIST);
+						if (pkgs != null) {
```
Comment (agatx) [inline zello-sdk/src/com/zello/sdk/Sdk.java]: Early return can be used here as well

File: zello-sdk/src/com/zello/sdk/Sdk.java [hunk 17]
```diff
@@ -81,33 +84,35 @@ void onCreate(String packageName, Context context) {
...
+							for (String pkg : pkgs) {
+								if (pkg.equalsIgnoreCase(_package)) {
+									reconnect();
+									updateSelectedContact(null);
+									updateContacts();
+									break;
```
Comment (agatx) [inline zello-sdk/src/com/zello/sdk/Sdk.java]: `return` would be more descriptive here because that's exit point for this method

**Comment by adambailey-** [comment]
Also, need to update the readme to make it say the lowest supported version is API 15. Both the .md and PDF

File: zello-sdk-sample-contacts/src/main/java/com/zello/sdk/sample/contacts/ContactsActivity.java [hunk 18]
```diff
@@ -1,211 +1,205 @@
...
+
+		// Use with an app installed from a generic PTT SDK apk obtained from https://github.com/zelloptt/zello-android-client-sdk/releases
+		Zello.getInstance().configure("com.pttsdk", this, this);
+
+		// Contact list pick handler
+		_contactsListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
+			@Override
+			public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
+				ListAdapter adapter = (ListAdapter) _contactsListView.getAdapter();
+				if (adapter != null) {
```
[the remark this answers is not in the document]
Comment (zello-eugene) [reply zello-sdk-sample-contacts/src/main/java/com/zello/sdk/sample/contacts/ContactsActivity.java]: Sure

File: zello-sdk-sample-contacts/src/main/java/com/zello/sdk/sample/contacts/ContactsActivity.java [hunk 19]
```diff
@@ -1,211 +1,205 @@
...
+	}
+
+	@Override
+	protected void onPause() {
+		super.onPause();
+		Zello.getInstance().enterPowerSavingMode();
+	}
+
+	@Override
+	public boolean onCreateOptionsMenu(Menu menu) {
```
[the remark this answers is not in the document]
Comment (zello-eugene) [reply zello-sdk-sample-contacts/src/main/java/com/zello/sdk/sample/contacts/ContactsActivity.java]: You betcha

File: zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java [hunk 20]
```diff
@@ -16,200 +16,201 @@
...
+		return true;
+	}
+
+	@Override
+	public boolean onOptionsItemSelected(MenuItem item) {
+		switch (item.getItemId()) {
+			case R.id.menu_available:
+			case R.id.menu_solo:
+			case R.id.menu_busy: {
```
[the remark this answers is not in the document]
Comment (zello-eugene) [reply zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java]: Of course

File: zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java [hunk 21]
```diff
@@ -16,200 +16,201 @@
...
-            updateUIForSignedIn();
-        } else if (_appState.isSigningIn()) {
-            updateUIForSigningIn();
-        } else {
-            updateUIForDisconnected();
-        }
-    }
+	@Override
+	public void onAudioStateChanged() {
+
```
[the remark this answers is not in the document]
Comment (zello-eugene) [reply zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java]: Naturally

File: zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java (hunk 1 above)
Comment (zello-eugene) [reply zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java]: You got it

File: zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java (hunk 2 above)
Comment (zello-eugene) [reply zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java]: Done

File: zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java (hunk 3 above)
Comment (zello-eugene) [reply zello-sdk-sample-misc/src/main/java/com/zello/sdk/sample/misc/MiscActivity.java]: Not necessary

File: zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java (hunk 5 above)
Comment (zello-eugene) [reply zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java]: Call me old fashioned, but I think this takes the rule too far.

File: zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java (hunk 4 above)
Comment (zello-eugene) [reply zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java]: Okay

File: zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java (hunk 6 above)
Comment (zello-eugene) [reply zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java]: Right

File: zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java (hunk 7 above)
Comment (zello-eugene) [reply zello-sdk-sample-ptt/src/main/java/com/zello/sdk/sample/ptt/PttActivity.java]: Fair

File: zello-sdk-sample-signin/src/main/res/values/strings.xml (hunk 8 above)
Comment (zello-eugene) [reply zello-sdk-sample-signin/src/main/res/values/strings.xml]: Cool

File: zello-sdk-sample/build.gradle (hunk 9 above)
Comment (zello-eugene) [reply zello-sdk-sample/build.gradle]: Yes

File: zello-sdk-sample/src/main/java/com/zello/sdk/sample/Helper.java (hunk 10 above)
Comment (zello-eugene) [reply zello-sdk-sample/src/main/java/com/zello/sdk/sample/Helper.java]: Nice

File: zello-sdk-sample/src/main/java/com/zello/sdk/sample/TalkActivity.java (hunk 11 above)
Comment (zello-eugene) [reply zello-sdk-sample/src/main/java/com/zello/sdk/sample/TalkActivity.java]: Thanks

File: zello-sdk/src/com/zello/sdk/Sdk.java (hunk 13 above)
Comment (zello-eugene) [reply zello-sdk/src/com/zello/sdk/Sdk.java]: \_O_/

File: zello-sdk/src/com/zello/sdk/Sdk.java (hunk 14 above)
Comment (zello-eugene) [reply zello-sdk/src/com/zello/sdk/Sdk.java]: Sure

File: zello-sdk/src/com/zello/sdk/Sdk.java (hunk 15 above)
Comment (zello-eugene) [reply zello-sdk/src/com/zello/sdk/Sdk.java]: Done

File: zello-sdk/src/com/zello/sdk/Sdk.java (hunk 16 above)
Comment (zello-eugene) [reply zello-sdk/src/com/zello/sdk/Sdk.java]: Yes

File: zello-sdk/src/com/zello/sdk/Sdk.java (hunk 17 above)
Comment (zello-eugene) [reply zello-sdk/src/com/zello/sdk/Sdk.java]: And yes

**Pull request merged**